Add taxonomy support to Product Collection Data API endpoint

dinhtungdu opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR adds taxonomy support to the Product Collection Data API endpoint, enabling clients to retrieve taxonomy counts (such as categories and tags) for products in a collection.

**Key changes:**
- Added `calculate_taxonomy_counts` parameter to the product collection data endpoint
- Implemented taxonomy count calculation logic that returns term IDs and their corresponding product counts
- Enhanced test coverage to verify taxonomy counts for both categories and tags

Closes WOOPLUG-4766
Closes https://github.com/woocommerce/woocommerce/issues/59130

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Set up a test environment with WooCommerce installed
2. Create test products and assign them to various categories and tags
3. Make a REST API request to `/wc/store/v1/products/collection-data` with the `calculate_taxonomy_counts` parameter:
   ```
   GET /wc/store/v1/products/collection-data?calculate_taxonomy_counts[]=product_cat&calculate_taxonomy_counts[]=product_tag
   ```
4. Verify the response includes a `taxonomy_counts` array with term IDs and their corresponding product counts
5. Test with different taxonomy combinations to ensure accurate count calculations
6. Run the PHPUnit tests to verify all functionality works correctly:
   ```
   pnpm run test:php:env tests/php/src/Blocks/StoreApi/Routes/ProductCollectionData.php
   ```

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message

Add taxonomy support to Product Collection Data API endpoint with calculate_taxonomy_counts parameter

</details>

dinhtungdu committed:
feat: add calculate_taxonomy_counts parameter to ProductCollectionData endpoint

dinhtungdu committed:
feat: implement taxonomy count calculation logic in ProductCollectionData

dinhtungdu committed:
feat: implement QueryClausesGenerator in ProductQuery and use FilterData for taxonomy counts

dinhtungdu committed:
feat: add taxonomy_counts field to ProductCollectionData schema

dinhtungdu committed:
docs: update ProductCollectionData API documentation for taxonomy support

dinhtungdu committed:
test: add taxonomy support tests to ProductCollectionData route

dinhtungdu committed:
fix: preserve term IDs in taxonomy counts

Use + operator instead of array_merge to preserve term IDs when
combining taxonomy counts from multiple taxonomies.

dinhtungdu committed:
refactor: simplify taxonomy counts parameter structure

Change calculate_taxonomy_counts from array of objects to simple
array of taxonomy name strings since query types are not needed
for taxonomies.

Before: calculate_taxonomy_counts[0][taxonomy]=product_cat
After: calculate_taxonomy_counts[0]=product_cat

dinhtungdu committed:
test: add count verification for taxonomy counts in multiple taxonomies test

dinhtungdu committed:
style: fix cs

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

gigitux commented on plugins/woocommerce/src/StoreApi/Utilities/ProductQueryFilters.php:
```diff
@@ -213,4 +213,39 @@ public function get_rating_counts( $request ) {
+		$query_vars    = $product_query->prepare_objects_query( $request );
+
+		// Use FilterData with ProductQuery as QueryClausesGenerator.
+		$container = wc_get_container();
```
Checking the documentation, in the src folder, the DI should be done in a different way: https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/README.md#1-other-classes-in-the-src-directory

dinhtungdu replied:
Store API (and Blocks) has its own container, so it's not possible to follow that guide right now. See https://github.com/woocommerce/woocommerce/pull/43772#discussion_r1478117593

gigitux commented on plugins/woocommerce/tests/php/src/Blocks/StoreApi/Routes/ProductCollectionData.php:
```diff
@@ -171,6 +176,107 @@ public function test_calculate_rating_counts() {
+	 * Test taxonomy calculation method.
+	 */
+	public function test_calculate_taxonomy_counts() {
+		$fixtures = new FixtureData();
```
It looks like not necessary

dinhtungdu replied:
Right, I will delete this. It was used initially but I forgot to remove after a refactor.

gigitux reviewed (changes requested):
Thanks for working on this! I left one comment regarding the DI in the `src` folder. Besides this, everything LGTM! I'm not approving because I want to ensure that I tested the PR correctly. 

`/wc/store/v1/products/collection-data?calculate_taxonomy_counts[]=product_cat&calculate_taxonomy_counts[]=product_tag`

Can you clarify this step in the testing instruction? If I use this URL structure, I got:

> Fatal error: Allowed memory size of 268435456 bytes exhausted (tried to allocate 20480 bytes)

With this URL, everything works as expected:

`/wc/store/v1/products/collection-data?calculate_taxonomy_counts=product_cat, product_tag` right?

dinhtungdu commented:
@gigitux both URLs should work fine. I'm curious what can cause the error for you. It's working as expected on my end. You should still get the same array of two either using comma separated string or the array notion. 🤔 

<img width="1269" height="696" alt="image" src="https://github.com/user-attachments/assets/2e613e5f-44ae-429f-b0a4-1493d76c6868" />
<img width="1269" height="696" alt="image" src="https://github.com/user-attachments/assets/15f1dbd8-a5f1-49c1-bb9c-776c9dc7b1f4" />

dinhtungdu committed:
test: remove unused fixture data

gigitux reviewed (approved):
Thanks for addressing the feedback!

It was an issue on my env because I tested again, and everything works as expected with both URL structures.ì 🚀

<img width="1895" height="1428" alt="Image" src="https://github.com/user-attachments/assets/433a5fa7-f6d7-4e61-bc4f-d984b851c1fa" />

dinhtungdu committed:
docs: update calculate_taxonomy_counts examples to use comma-separated syntax

- Change from array syntax calculate_taxonomy_counts[0]=product_cat to comma-separated calculate_taxonomy_counts=product_cat
- Update cURL example to show multiple taxonomies with comma separation
- Both formats work due to WordPress REST API parameter parsing, but comma-separated is more concise
